File: Backtracking/DFS/39CombinationSum.py
from typing import List

# 经典DFS
# 这里需要传Start是为了防止重复. 如果从第一个出发, 它将找到所有和第一个有关的情况, 那么从第二个出发的时候就不能涵盖第一个了

class Solution:
    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
        list = []
        res = []
        def dfs(path, target, start):
            if target == 0:
                res.append(path)
                return
            elif target < 0:
                return
            
            for i in range(start, len(candidates)):
                dfs(path+[candidates[i]], target-candidates[i], i)
        
        dfs(list, target, 0)
        return res

File: Backtracking/DFS/39CombinationSum.java
import java.util.ArrayList;
import java.util.List;

class Solution {
    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        List<List<Integer>> res = new ArrayList<>();
        dfs(candidates, target, res, new ArrayList<>(), 0);
        return res;
    }

    private void dfs(int[] candidates, int target, List<List<Integer>> res, List<Integer> cur, int index) {

        if (target == 0) {
            res.add(new ArrayList<>(cur));
            return;
        }

        if (target < 0) {
            return;
        }

        for (int i = index; i < candidates.length; i++) {
            cur.add(candidates[i]);
            dfs(candidates, target - candidates[i], res, cur, i);
            cur.remove(cur.size() - 1);
        }
    }
}